core/shared/platform/common/posix/posix_thread.c

```diff
@@ -509,8 +509,8 @@ mask_signals(int how)
     pthread_sigmask(how, &set, NULL);
 }
 
-static os_thread_local_attribute struct sigaction prev_sig_act_SIGSEGV;
-static os_thread_local_attribute struct sigaction prev_sig_act_SIGBUS;
+static struct sigaction prev_sig_act_SIGSEGV;
```

Comment:
Could you try keeping the os_thread_local_attribute flag and call `wasm_runtime_init_thread_env` in the callback of the spawned thread? In that way prev_sig_act_SIGSEGV of that thread will be set.

Per my understanding, each thread can has its sig action, we had better keep the os_thread_local_attribute flag.

Reply:
My understanding is that the sigaction is per process, not per thread:
> The sigaction() system call is used to change the action taken by a process on receipt of a specific signal. 

(source: https://man7.org/linux/man-pages/man2/sigaction.2.html)

Reply:
This is the answer I got from ChapGPT:
<img width="542" alt="image" src="https://github.com/bytecodealliance/wasm-micro-runtime/assets/20393758/f86e82d2-4611-453c-97cc-a9f69184379e">

Not very sure, but if keeping os_thread_local_attribute flag works for you, had better keep it.

Reply:
I don't think that's very reliable.
>The signal disposition is a per-process attribute: in a
       multithreaded application, the disposition of a particular signal
       is the same for all threads.

(https://man7.org/linux/man-pages/man7/signal.7.html)
From what I read, signal handers are per process, signal masks are per thread.

If I keep `os_thread_local_attribute`, it doesn't work. And I can't use `wasm_runtime_init_thread_env` in that thread since it doesn't import wamr dependencies. 
I don't think that's any harm in having the prev sigaction global instead of thread-local.

Reply:
OK, let's merge it.

Reply:
@wenyongh do you remember why you added sigaction call here?

Reply:
well, it seems that i replied to a wrong thread.
my question is about the sigaction call in signal_callback.

Reply:
@yamt Someone raised issue that previous signal handlers were supressed: https://github.com/bytecodealliance/wasm-micro-runtime/issues/1262
And I submitted PR https://github.com/bytecodealliance/wasm-micro-runtime/pull/1268 to fix it.

Reply:
> @yamt Someone raised issue that previous signal handlers were supressed: #1262 And I submitted PR #1268 to fix it.

i can understand the intention to call prev_sig_act->sa_sigaction.
but i don't understand why to call sigaction.

Reply:
Do you mean the handling of SIG_DFL? I am a little confused also, I referred to the implementation of wasmtime:
https://github.com/bytecodealliance/wasmtime/blob/main/crates/runtime/src/traphandlers/unix.rs#L169-L170:
```rust
    else if previous.sa_sigaction == libc::SIG_DFL || previous.sa_sigaction == libc::SIG_IGN {
        libc::sigaction(signum, previous, ptr::null_mut());
    }
```
But seems that its implementation is invalid: should use `previous.sa_handler` instead of `previous.sa_sigaction`, and should handle SIG_DFL and SIG_IGN in different ways? Also not sure whether using libc::sigaction is right.

In IBM's document: https://www.ibm.com/docs/en/zos/2.4.0?topic=functions-sigaction-examine-change-signal-action
```C
void (*)(int, siginfo_t *, void *) sa_sigaction
A pointer to the function assigned to handle the signal, or SIG_DFL, or SIG_IGN. 
```
And it mentions:
```C
Notes:
The user must cast SIG_IGN or SIG_DFL to match the sa_sigaction definition. (indicating that the signal is to be ignored).
```

Reply:
the intention of the wasmtime code seems like restore the signal handler and then make it re-fault.
i'm not sure it it works reliably.
even if it does, it leaves the signal handler modified, which seems broken for our purpose.
maybe it might not be a problem for wasmtime, where "multi-tenancy" is somehow optional.

Reply:
So any suggestion to handle SIG_DFL here? Or keep the code unchanged?

Reply:
my suggestion is to remove the SIG_DFL (and probably SIG_IGN) cases because apps can hardly rely on them.

Reply:
Not sure about those cases to be honest, personally I only tested the `prev_sig_act->sa_sigaction` and `prev_sig_act->sa_handler` cases.

Reply:
OK, so let's just handle the two cases, I will submit another PR to fix it.